persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java

```diff
@@ -86,17 +136,57 @@
     if (options == null) {
       return org.apache.cassandra.cql3.QueryOptions.DEFAULT;
     }
+
     org.apache.cassandra.transport.ProtocolVersion protocolVersion =
         toInternal(options.getProtocolVersion());
-    // TODO(mpenick): timestamp and nowInSeconds?
-    return org.apache.cassandra.cql3.QueryOptions.create(
+    return createOptions(
         toInternal(options.getConsistency()),
         options.getValues(),
+        options.getNames(),
         options.skipMetadata(),
         options.getPageSize(),
-        PagingState.deserialize(options.getPagingState(), protocolVersion),
+        // Note that PagingState.deserialize modify its input, so we duplicate to avoid nasty
+        // surprises down the line
+        PagingState.deserialize(options.getPagingState().duplicate(), protocolVersion),
         toInternal(options.getSerialConsistency()),
-        protocolVersion);
+        protocolVersion,
+        options.getTimestamp());
+  }
+
+  private static org.apache.cassandra.cql3.QueryOptions createOptions(
+      org.apache.cassandra.db.ConsistencyLevel consistencyLevel,
+      List<ByteBuffer> values,
+      List<String> boundNames,
+      boolean skipMetadata,
+      int pageSize,
+      PagingState pagingState,
+      org.apache.cassandra.db.ConsistencyLevel serialConsistency,
+      org.apache.cassandra.transport.ProtocolVersion protocolVersion,
+      long timestamp) {
+
+    org.apache.cassandra.cql3.QueryOptions options;
+    try {
+      Object specificOptions =
+          specificOptionsCtor.newInstance(pageSize, pagingState, serialConsistency, timestamp);
+      Object defaultOptions =
+          defaultOptionsCtor.newInstance(
+              consistencyLevel, values, skipMetadata, specificOptions, protocolVersion);
+      options = (org.apache.cassandra.cql3.QueryOptions) defaultOptions;
+
+      // Adds names if there is some.
+      if (boundNames != null) {
+        options =
+            (org.apache.cassandra.cql3.QueryOptions)
+                optionsWithNameCtor.newInstance(defaultOptions, boundNames);
+      }
+    } catch (Exception e) {
+      // We can't afford to ignore that: the values wouldn't be in the proper order, and worst
+      // case scenario, this could end up inserting values in the wrong columns, which essentially
+      // boils down to corrupting the use DB.
+      throw new RuntimeException(
```

Comment:
Should we throw a `ServerError` here?

Reply:
Maybe? I expected that we catch all exceptions somewhere and transform unknown ones into `ServerError` anyway. That's kind of how it works in C*: `ServerError` means "any exception that we didn't really expected to be throw but there is a bug", but we don't go and throw `ServerError` all over the place.